File: src/main/kotlin/ai/astha/scantheplanet/idea/scanner/providers/LocalModel.kt
package ai.astha.scantheplanet.idea.scanner.providers

import ai.astha.scantheplanet.idea.scanner.evidence.EvidenceParser
import ai.astha.scantheplanet.idea.scanner.model.CodeModel
import ai.astha.scantheplanet.idea.scanner.model.ModelFinding
import ai.astha.scantheplanet.idea.scanner.prompt.BatchPromptPayload
import ai.astha.scantheplanet.idea.scanner.prompt.PromptPayload

class LocalModel(private val severityOverride: String? = null) : CodeModel {
    override val name: String = "local"

    override fun analyzeChunk(prompt: PromptPayload): List<ModelFinding> {
        if (prompt.codeChunk.ruleHints.isEmpty()) return emptyList()
        val severity = severityOverride?.lowercase() ?: "info"
        return prompt.codeChunk.ruleHints.map { hint ->
            ModelFinding(
                techniqueId = null,
                chunkId = prompt.codeChunk.id,
                file = prompt.codeChunk.file,
                startLine = hint.line,
                endLine = hint.line,
                severity = severity,
                confidence = null,
                observation = "Rule hint matched: ${hint.signalId}",
                evidence = EvidenceParser.formatRuleHintEvidence(prompt.codeChunk.file, hint.line, hint.snippet),
                reasoning = null,
                unknownMitigations = emptyList(),
                modelName = name
            )
        }
    }

    override fun analyzeChunkBatch(prompt: BatchPromptPayload): List<ModelFinding> {
        val severity = severityOverride?.lowercase() ?: "info"
        val findings = mutableListOf<ModelFinding>()
        for (technique in prompt.techniques) {
            if (technique.ruleHints.isEmpty()) continue
            for (hint in technique.ruleHints) {
                findings.add(
                    ModelFinding(
                        techniqueId = technique.id,
                        chunkId = prompt.codeChunk.id,
                        file = prompt.codeChunk.file,
                        startLine = hint.line,
                        endLine = hint.line,
                        severity = severity,
                        confidence = null,
                        observation = "Rule hint matched: ${hint.signalId}",
                        evidence = EvidenceParser.formatRuleHintEvidence(prompt.codeChunk.file, hint.line, hint.snippet),
                        reasoning = null,
                        unknownMitigations = emptyList(),
                        modelName = name
                    )
                )
            }
        }
        return findings
    }
}

File: packages/core/src/providers/LocalModel.ts
import type { CodeModel } from './CodeModel.js';
import type { ModelFinding } from '../types/Finding.js';
import type { PromptPayload, BatchPromptPayload } from '../prompt/PromptModels.js';
import { formatRuleHintEvidence } from './EvidenceParser.js';

/**
 * Local model that uses heuristic matching only (no LLM calls)
 */
export class LocalModel implements CodeModel {
  readonly name = 'local';

  constructor(private readonly severityOverride?: string) {}

  async analyzeChunk(prompt: PromptPayload): Promise<ModelFinding[]> {
    if (prompt.codeChunk.ruleHints.length === 0) {
      return [];
    }

    const severity = this.severityOverride?.toLowerCase() ?? 'info';

    return prompt.codeChunk.ruleHints.map((hint) => ({
      techniqueId: undefined,
      chunkId: prompt.codeChunk.id,
      file: prompt.codeChunk.file,
      startLine: hint.line,
      endLine: hint.line,
      severity,
      confidence: null,
      observation: `Rule hint matched: ${hint.signalId}`,
      evidence: formatRuleHintEvidence(
        prompt.codeChunk.file,
        hint.line,
        hint.snippet
      ),
      reasoning: null,
      unknownMitigations: [],
      modelName: this.name,
    }));
  }

  async analyzeChunkBatch(prompt: BatchPromptPayload): Promise<ModelFinding[]> {
    const severity = this.severityOverride?.toLowerCase() ?? 'info';
    const findings: ModelFinding[] = [];

    for (const technique of prompt.techniques) {
      if (technique.ruleHints.length === 0) continue;

      for (const hint of technique.ruleHints) {
        findings.push({
          techniqueId: technique.id,
          chunkId: prompt.codeChunk.id,
          file: prompt.codeChunk.file,
          startLine: hint.line,
          endLine: hint.line,
          severity,
          confidence: null,
          observation: `Rule hint matched: ${hint.signalId}`,
          evidence: formatRuleHintEvidence(
            prompt.codeChunk.file,
            hint.line,
            hint.snippet
          ),
          reasoning: null,
          unknownMitigations: [],
          modelName: this.name,
        });
      }
    }

    return findings;
  }
}
